pyquil/api/_qam.py

```diff
@@ -60,18 +61,27 @@ def load(self, executable):
         return self
 
     @_record_call
-    def write_memory(self, *, region_name: str, offset: int = 0, value=None):
+    def write_memory(self, *, region_name: str, offset: int = None,
+                     value: Optional[Union[int, float, Sequence[int], Sequence[float]]] = None):
         """
-        Writes a value into a memory region on the QAM at a specified offset.
+        Writes a value or chronologically unwraps a list of values into a memory region on
+        the QAM at a specified offset.
 
         :param region_name: Name of the declared memory region on the QAM.
         :param offset: Integer offset into the memory region to write to.
-        :param value: Value to store at the indicated location.
+        :param value: Value(s) to store at the indicated location.
         """
         assert self.status in ['loaded', 'done']
 
-        aref = ParameterAref(name=region_name, index=offset)
-        self._variables_shim[aref] = value
+        if isinstance(value, Sequence):
```

Comment:
The original version did no type checking, but I would prefer to have some checks here. For example, the region indicated by `region_name` presumably has some declared type, so we might as well check that the provided values are of this type. If there is a disagreement between the provided value and the region type, perhaps the pythonic thing to do here is cast (e.g. `int` to `float` for a `REAL` region). An alternative, less stringent check (in case you do not have access to declared region types here) might just be to ensure that in the case of a value `Sequence`, that the `Sequence` elements have the same type.

Reply:
Also might check that the sequence length does not exceed the length of the memory region.

Reply:
Doesn't the fact that we already specified the `region_name` parameter type in the declaration mean we don't need to do an additional assertion of its type? And if not, is an `assert` statement the best way to do that?

Reply:
Unfortunately, all but the most explicit type checks in python are enforced "very optionally" (read: not at all). An `assert` will do; if you write one, make sure to use of `assert`'s optional argument where you write a human-readable message about what you were expecting / what went wrong.